main/src/ca/uwaterloo/flix/language/phase/LambdaLift.scala

```diff
@@ -360,6 +360,14 @@ object LambdaLift extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
         val e2 = visitExp(exp2)
         Expression.FixpointEntails(e1, e2, tpe, loc)
 
+      case Expression.FixpointFold(pred, exp1, exp2, exp3, tpe, loc) =>
+        val p = visitPredicateWithParam(pred)
+        val e1 = visitExp(exp1).asInstanceOf[Expression.Var] // TODO: not really clean
```

Comment:
Wait, why does this cast have to happen in multiple places. Ideally I would just have it in one phase at some appropriate point in the pipeline.

Reply:
The problem is that `visitExp` returns an `Expression`, while `Expression.FixpointFold` expects `e1` to be an `Expression.Var`. This could be solved by just having `Expression.FixpointFold` take a more generic `Expression`, and restricting it to an `Expression.Var` only in the last AST transformation, but this could force the casts to happen in other places (e.g., the typing phase)

Reply:
Can we not just change the type of FixpointFold to accept expressions, and only let it have variables in FinalAst? It is less type sound, but more robust for a future refactoring.